### PR TITLE
Adjust batch mode partition count calculation

### PR DESCRIPTION
**Pull request by kastiglione**
While exploring how to optimize global build times, I noticed that the batch sizing logic wasn't producing multiple batches until a module had 50 files. Based on @graydon's [detailed batch calculation comment](https://github.com/apple/swift/blob/746b58e8e1c1ef1ac09d8031875fd2a08b65597c/lib/Driver/Compilation.cpp#L1269-L1362) (which says "cap the batch size ... at ... 25"), and based on the variable names ("`SizeLimit`"), I believe the batch calculation implementation is off.

The simple example is a module of 26 files. With the existing batch calculation, the result is a `26/25=1` batch. However this is larger than the `DefaultSizeLimit`. This calculation continues to yield `1` for modules up to 49 files. Only at modules of 50 files does the batch calculation produce two batches.

Here's a table showing the effects of the current partition count calculation:

| Num Files | Num Batches  | Min Batch Size | Max Batch Size |
| --- | --- | --- | --- |
| 1-49 | 1 | 1 | 49 |
| 50-74 | 2 | 25 | 37 |
| 75-99 | 3 | 25 | 33 |
| 100-124 | 4 | 25 | 31 |

This table shows that the max batch size is inversely proportional with the number of files. Smaller modules have a higher max batch size, and as a module gets larger, its max batch size trends toward 25. It also shows that after 50, the minimum batch size is consistently 25.

With this proposed change to the batch count calculation, the math is now:

| Num Files | Num Batches  | Min Batch Size | Max Batch Size |
| --- | --- | --- | --- |
| 1-25 | 1 | 1 | 25 |
| 26-50 | 2 | 13 | 25 |
| 51-75 | 3 | 17 | 25 |
| 76-100 | 4 | 19 | 25 |
| 101-125 | 5 | 21 | 25 |

This table shows that the batch size is in fact capped to 25, independent of how many files are in the module. It also show the minimum batch size increases with module size.

### CODE REVIEW

**Comment by kastiglione** [comment]
I'm presuming this was never noticed because Xcode's use of `-j` means it doesn't rely on this logic. We use Bazel which doesn't inject a `-j` value for you.

File: lib/Driver/Compilation.cpp
Comment (davidungar) [inline lib/Driver/Compilation.cpp]: Two comments:
1. Suggest you define a lambda for the round-up, or better yet, use a library function. Either way would make things a big clearer. (Your comment is great--keep it!)
2. I don't think (x - 1) / y + 1 is the canonical formula. The one I recall is (x + y - 1) / y. Your formula is equivalent, but it's too easy to read it as "(x - 1) / (y + 1)", which is how I read it. But best would be to just use existing library routines.

**Review by davidungar — CHANGES_REQUESTED** [review]
Thanks for catching this issue, it is a good catch! You'll see in my comment at the change that I'd like to ask for a minor change to the comment (if you agree). My more critical ask is that you use a library routine for the round up. I'm sorry to ask for more work, but in a code base this large, my feeling is anything we can do to help out a future maintainer is worth while. Using a library routine will obviate the need for a future maintainer to double-check your math. Your comment is great! Thanks.

File: lib/Driver/Compilation.cpp
Comment (kastiglione) [reply lib/Driver/Compilation.cpp]: Thanks I'll look to see if there's an existing function. I too considered `(x + y - 1) / y`, and for whatever reason I ended up this way. I'm happy to change it.

Comment (kastiglione) [reply lib/Driver/Compilation.cpp]: Your forumla also handles x = 0.

Comment (davidungar) [reply lib/Driver/Compilation.cpp]: My pleasure. Good point!

**Comment by DougGregor** [comment]
We should mirror this change in [swift-driver](https://github.com/apple/swift-driver) as well

**Comment by kastiglione** [comment]
@DougGregor I've opened a PR with the mirrored change: https://github.com/apple/swift-driver/pull/51.

**Review by davidungar — COMMENTED** [review]
A library function would be better, but this is ok to merge if need be.

**Review by davidungar — APPROVED** [review]

**Comment by kastiglione** [comment]
I'm happy to move it, do you have any suggestions for where? I first looked to make it an addition to an existing library, but I didn't see any headers that looked right (looking mostly in Basic). I didn't feel I should take the step to start one, so I went with the last choice, a lambda as you had suggested.

**Comment by davidungar** [comment]
> I'm happy to move it, do you have any suggestions for where? I first looked to make it an addition to an existing library, but I didn't see any headers that looked right (looking mostly in Basic). I didn't feel I should take the step to start one, so I went with the last choice, a lambda as you had suggested.

Thanks for looking for a library function. I looked very briefly and didn't find one either, sigh. The current version is fine, modulo one tiny little nit: `DivideUp` could be confused for having to do with dividing up the jobs into partitions. `DivideRoundingUp` or `CeilingOfQuotient` (bleh) or some such might be a tad better. Feel free to ignore this nit, though.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by DougGregor** [comment]
Thanks! I'll go ahead and merge this. @kastiglione, if you cherry-pick to the `swift-5.2-branch`, we can merge there, too.

**Comment by kastiglione** [comment]
Cherry-pick here https://github.com/apple/swift/pull/29021

**Pull request merged**
